### PR TITLE
Orca Pytorch Estimator Tensorboard

### PR DESCRIPTION
**Pull request by jing-xu**
This is an example ploting training loss vs. epochs. Before running the code, you should have tensorboardX installed by "pip install tensorboardX". Then you could run  "tensorboard --logdir=runs" and see plots showing on http://localhost:6006. This code has also been tested on Almaren Node 002 with 2 nodes.

### CODE REVIEW

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+from __future__ import print_function
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:1]: license header

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    for stat in stats:
+        writer.add_scalar("training_loss", stat['train_loss'], stat['epoch'])
+    
+    print("Train stats: {}".format(stats))
+    val_stats = zoo_estimator.evaluate(validation_data_creator)
+    print("validation stats: {}".format(val_stats))
+    stop_orca_context()
+
+if __name__ == '__main__':
+    main()
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:115]: new line

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import torch
+import torchvision
+import torchvision.transforms as transforms
+
+import numpy as np
+
+import torch.nn as nn
+import torch.nn.functional as F
+
+import torch.optim as optim
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:14]: format the imports (i.e. put torch related together, generic imports together)

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+from tensorboardX import SummaryWriter
+
+from zoo.orca import init_orca_context, stop_orca_context
+from zoo.orca.learn.pytorch import Estimator
+from zoo.orca.learn.metrics import Accuracy
+from zoo.orca.learn.trigger import EveryEpoch
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:21]: not used

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    trainloader = torch.utils.data.DataLoader(trainset, batch_size=config.get("batch_size",32),
+                                              shuffle=True, num_workers=2)
+    return trainloader
+    
+def validation_data_creator(config):
+    transform = transforms.Compose(
+        [transforms.ToTensor(),
+        transforms.Normalize((0.5,), (0.5,))])
+    testset = torchvision.datasets.FashionMNIST(root='./data', train=False,
+                                           download=True, transform=transform)
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:68]: align with the previous line

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    testloader = torch.utils.data.DataLoader(testset, batch_size=config.get("batch_size",32),
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:69]: space after ,

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                        help="The number of workers to run on each node")
+    parser.add_argument('--epochs', type=int, default=2, metavar='N',
+                        help='number of epochs to train (default: 2)')
+    parser.add_argument('--worker_batch', type=int, default=16, metavar='N',
+                        help='input worker batch for training per executor(default: 16)')
+    args = parser.parse_args()
+    
+    init_orca_context(cluster_mode=args.cluster_mode, cores=args.cores, num_nodes=args.num_nodes, memory=args.memory)
+    
+    classes = ('T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:98]: is it used?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle Boot')
+    criterion = nn.CrossEntropyLoss()
+    zoo_estimator = Estimator.from_torch(model=model_creator, optimizer=optimizer_creator, loss=criterion, config={
+        "batch_size":args.worker_batch}, backend="pytorch")
+    stats = zoo_estimator.fit(train_data_creator, epochs=args.epochs)
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:103]: move batch size to fit

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import torch.nn.functional as F
+import torch.optim as optim
+
+from tensorboardX import SummaryWriter
+
+from zoo.orca import init_orca_context, stop_orca_context
+from zoo.orca.learn.pytorch import Estimator
+
+
+def train_data_creator(config={}):
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:39]: just config; also modify for other creators

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    transform = transforms.Compose(
+        [transforms.ToTensor(),
+        transforms.Normalize((0.5,), (0.5,))])
+    testset = torchvision.datasets.FashionMNIST(root='./data', train=False,
+                                                download=True, transform=transform)
+    testloader = torch.utils.data.DataLoader(testset, batch_size=4,
+                                             shuffle=False, num_workers=2)
+    return testloader
+
+def matplotlib_imshow(img, one_channel=False):
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:62]: Add this comment in the original tutorial: `# helper function to show an image`

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        init_orca_context(memory="4g")
+    elif args.cluster_mode == "yarn":
+        init_orca_context(cluster_mode=args.cluster_mode, cores=4, num_nodes=2, memory="4g")
+
+    writer = SummaryWriter('runs/fashion_mnist_experiment_1')
+    # constant for classes
+    classes = ('T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
+            'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle Boot')
+
+    # plot some random training images
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:115]: # get some random training images
# create grid of images
# show images
Keep these comments in the original tutorial?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    img_grid = torchvision.utils.make_grid(images)
+    matplotlib_imshow(img_grid, one_channel=True)
+
+    # write to tensorboard
+    writer.add_image('four_fashion_mnist_images', img_grid)
+
+    # training loss vs. epochs
+    criterion = nn.CrossEntropyLoss()
+    orca_estimator = Estimator.from_torch(model=model_creator, optimizer=optimizer_creator, loss=criterion, config={}, backend="torch_distributed")
+    stats = orca_estimator.fit(train_data_creator, epochs=5, batch_size=4)
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:127]: the tutorial seems only train for 1 epoch right?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+def train_data_creator(config):
+    transform = transforms.Compose(
+        [transforms.ToTensor(),
+        transforms.Normalize((0.5,), (0.5,))])
+
+    trainset = torchvision.datasets.FashionMNIST('./data',
+        download=True,
+        train=True,
+        transform=transform)
+    trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/tensorboard.py:48]: batch_size=config["batch_size"]

**Comment by jing-xu** [comment]
This example shows how to use tensorboard with Orca. Three main functions in Pytorch tensorboard tutorial are adapted here: showing a batch of training images, inspecting the model and plotting loss vs epoch. The original tutorial plots loss vs training data in one epoch. Since we can't get intermediate training loss, this example changes to loss vs epoch, where epoch is set to 5.

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    # write to tensorboard
+    writer.add_image('four_fashion_mnist_images', img_grid)
+    
+    # inspect the model using tensorboard
+    writer.add_graph(model_creator(config={}), images)
+    writer.close()
+
+    # training loss vs. epochs
+    criterion = nn.CrossEntropyLoss()
+    orca_estimator = Estimator.from_torch(model=model_creator, optimizer=optimizer_creator, loss=criterion, config={"batch_size":4}, backend="torch_distributed")
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:135]: is this line too long to pass the style check?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    return optimizer
+
+def main():
+    parser = argparse.ArgumentParser(description='PyTorch Tensorboard Example')
+
+    parser.add_argument('--cluster_mode', type=str, default="local",
+                        help='The mode for the Spark cluster.')
+    args = parser.parse_args()
+    if args.cluster_mode == "local":
+        init_orca_context(memory="4g")
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:107]: does this need 4g memory?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    # training loss vs. epochs
+    criterion = nn.CrossEntropyLoss()
+    orca_estimator = Estimator.from_torch(model=model_creator, optimizer=optimizer_creator, loss=criterion, config={"batch_size":4}, backend="torch_distributed")
+    stats = orca_estimator.fit(train_data_creator, epochs=5, batch_size=4)
+
+    for stat in stats:
+        writer.add_scalar("training_loss", stat['train_loss'], stat['epoch'])
+    print("Train stats: {}".format(stats))
+    val_stats = orca_estimator.evaluate(validation_data_creator)
+    print("validation stats: {}".format(val_stats))
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:142]: v -> V

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    # inspect the model using tensorboard
+    writer.add_graph(model_creator(config={}), images)
+    writer.close()
+
+    # training loss vs. epochs
+    criterion = nn.CrossEntropyLoss()
+    orca_estimator = Estimator.from_torch(model=model_creator,
+                                          optimizer=optimizer_creator,
+                                          loss=criterion,
+                                          config={"batch_size": 4},
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:144]: do we need batch_size here? it is also specified in `fit` below?

**Comment by jason-dai** [comment]
> This example shows how to use tensorboard with Orca. Three main functions in Pytorch tensorboard tutorial are adapted here: showing a batch of training images, inspecting the model and plotting loss vs epoch. The original tutorial plots loss vs training data in one epoch. Since we can't get intermediate training loss, this example changes to loss vs epoch, where epoch is set to 5.

Shall we return loss for every epoch? @yangw1234 @hkvision

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+def optimizer_creator(model, config):
+    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
+    return optimizer
+
+
+def main():
+    parser = argparse.ArgumentParser(description='PyTorch Tensorboard Example')
+
+    parser.add_argument('--cluster_mode', type=str, default="local",
+                        help='The mode for the Spark cluster.')
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:110]: help='The cluster mode, such as local, yarn or k8s.'

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    args = parser.parse_args()
+    if args.cluster_mode == "local":
+        init_orca_context()
+    elif args.cluster_mode == "yarn":
+        init_orca_context(cluster_mode=args.cluster_mode, cores=4, num_nodes=2)
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:115]: cluster_mode needs to be "yarn-client" here, no?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py (hunk 17 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:144]: Ar, remove this here @jing-xu Only need to specify in fit.

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py (hunk 19 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/fashion-mnist.py:115]: yarn will also be mapped to yarn-client in init_orca_context

**Comment by hkvision** [comment]
> > This example shows how to use tensorboard with Orca. Three main functions in Pytorch tensorboard tutorial are adapted here: showing a batch of training images, inspecting the model and plotting loss vs epoch. The original tutorial plots loss vs training data in one epoch. Since we can't get intermediate training loss, this example changes to loss vs epoch, where epoch is set to 5.
> 
> Shall we return loss for every epoch? @yangw1234 @hkvision

We are already returning the loss for every epoch.
What she means is the original tutorial draws the graph loss vs every batch. The loss of each __batch__ is not returned in our implementation.

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
+# Pytorch Tensorboard example
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md:1]: Change folder name to fashion_mnist?

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+We demostrate how to easily show the graphical results of runing synchronous distributed Pytorch training using Pytorch Estimator of Project Orca in Analytics Zoo. We use a simple convolutional nueral network model to train on fashion-MNIST dataset. See [here](https://pytorch.org/tutorials/intermediate/tensorboard_tutorial.html) for the original single-node version of this example provided by Pytorch.
+
+## Prepare environments
+
+We recommend you to use Anaconda to prepare the environments, especially if you want to run on a yarn cluster
+
+```
+conda create -n zoo python=3.7 # zoo is conda environment name, you can set another name you like.
+conda activate zoo
+pip install analytics-zoo # 0.9.0.dev0 or above
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md:11]: analytics-zoo[ray]

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+pip install torch
+pip install torchvision
+pip install ray==0.8.4
+pip install tensorboardX
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md:15]: remove these two lines

File: pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+pip install matplotlib
+```
+
+# Run on local after pip install
+
+```
+python tensorboard.py
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/tensorboard/README.md:22]: wrong file name

**Comment by hkvision** [comment]
Temporarily disable the jenkins test due to the low torch version: https://github.com/intel-analytics/analytics-zoo/issues/3233

File: pyzoo/zoo/examples/orca/learn/pytorch/fashion-mnist/README.md [hunk 24]
```diff
@@ -0,0 +1,45 @@
+# Pytorch Fashion-MNIST example on Tensorboard
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/fashion-mnist/README.md:1]: on -> with Tensorboard for visualization

**Comment by jing-xu** [comment]
The torch_distributed backend has wrappers for train and validation data (train_data_creator and validation_data_creator), which need additional load to get trainset for the 'projector' part. And for 'pr_curve', since our estimator won't return intermediate outputs, we need to modify the implementation of pytorch_ray_estimator in order to calculate that.
